services/cleanup/models.py

```diff
@@ -0,0 +1,199 @@
+import dataclasses
+import itertools
+from collections import defaultdict
+from collections.abc import Callable
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+
+from django.db.models import Model
+from django.db.models.query import QuerySet
+from shared.bundle_analysis import StoragePaths
+from shared.django_apps.compare.models import CommitComparison
+from shared.django_apps.core.models import Commit, Pull
+from shared.django_apps.profiling.models import ProfilingUpload
+from shared.django_apps.reports.models import CommitReport, ReportDetails
+from shared.django_apps.reports.models import ReportSession as Upload
+from shared.django_apps.staticanalysis.models import StaticAnalysisSingleFileSnapshot
+
+from services.archive import ArchiveService, MinioEndpoints
+from services.cleanup.utils import CleanupContext, CleanupResult
+
+MANUAL_QUERY_CHUNKSIZE = 5_000
+DELETE_FILES_BATCHSIZE = 50
+
+
+def cleanup_files_batched(
+    context: CleanupContext, buckets_paths: dict[str, list[str]]
+) -> int:
+    cleaned_files = 0
+
+    # TODO: maybe reuse the executor across calls?
```

Comment:
We certainly stand to gainfrom not having to initialize and tear down the pool every time.

But it seems that the calls to `cleanup_files_batched` themselves are not concurrent. Is that on purpose? Like we are now it seems that at any given time there would always be at most 1 pool running anyway

Reply:
Indeed, it is running only with one pool at a time.

The regular deletions are still running serially in topological sort order. So it does not concurrently delete two unrelated models.

I was thinking about that a bit, but I thought it would be too complex to actually implement. Also reusing the thread pool would mean that I would have to introduce some kind of "waitpool" to make sure all the file deletes are complete before returning, which still keeping the pool alive.
I wanted to avoid that complexity for now.